games/xworld3d/tasks/xworld3d_task.py

```diff
@@ -0,0 +1,355 @@
+"""
+Copyright (c) 2017 Baidu Inc. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from context_free_grammar import CFG
+from py_gflags import get_flag
+
+class XWorld3DTask(object):
+    ## some static class variables
+    ## that shoule be shared by all derived classes
+    time_penalty = -0.1
+    correct_reward = 1.0
+    wrong_reward = -1.0
+    failed_action_penalty = -0.2
+
+    def __init__(self, env):
+        ## define all the spatial relations
+        self.directions = {
+            (1, 0) : "east",
+            (-1, 0) : "west",
+            (0, 1) : "south",
+            (0, -1) : "north",
+            (1, 1) : "southeast",
+            (1, -1) : "northeast",
+            (-1, 1) : "southwest",
+            (-1, -1) : "northwest"
+        }
+        self.distance_threshold = 1.0
+        self.orientation_threshold = 0.707
+        self.env = env
+        self.event = ""
+        self.num_successes = 0
+        self.num_failures = 0
+        self.reset(True)
+        self.cfg = CFG(*self._define_grammar())
+
+    ################ internal functions ####################
+    def _define_grammar(self):
+        """
+        The derived class can override this function to define a grammar
+        for the teacher. To do so, the function should return a tuple of a grammar
+        string and a start symbol.
+        By default the grammar will be empty and the teacher generates
+        a empty sentence each time step.
+        """
+        return "", ""
+
+    def _get_all_directions(self):
+        """
+        Return all the spatial-relation words in xworld.
+        """
+        return self.directions.values()
+
+    def _get_all_colors(self):
+        """
+        Return all colors defined in xworld.
+        """
+        return self.env.get_all_colors()
+
+    def _get_direction_and_distance(self, l1, l2):
+        """
+        Get the direciton of l2 wrt l1. For example, if the result is 'X'
+        then 'l2 is in the X of l1'.
+        When obtaining the direction, we use (l2 - l1)
+        """
+        dx = l2[0] - l1[0]
+        dy = l2[1] - l1[1]
+        best_sim = []
+        best_dir = []
+        dist = pow(pow(dx, 2) + pow(dy, 2), 0.5)
+        for d in self.directions.keys():
+            sim = (dx*d[0] + dy*d[1]) / pow(d[0]*d[0]+d[1]*d[1], 0.5)
```

Comment:
In first-person, the directions become "front", "behind", "left of", "right of". You can have "front-left", etc. But "north", "west" etc are absolute directions. It is really weird to have these words especially when the teacher refers the direction w.r.t. the agent itself.